gnomad/utils/constraint.py

```diff
@@ -877,3 +879,357 @@ def compute_expected_variants(
+    :return: Table with pLI, pNull, and pRec scores.
+    """
+    full_pli_expr = defaultdict(list)
+    # Calculate pLI scores for each population if specified.
+    if pops:
+        pop_lengths = get_all_pop_lengths(ht, pops, f"obs_{annotation_name}_")
+        for pop_length, pop in pop_lengths:
+            logger.info(f"Calculating pLI for {pop}...")
+            for i in range(n_ds_to_skip, pop_length):
+                logger.info(i)
+                pli_expr = compute_pli(
+                    ht,
+                    ht[f"obs_{annotation_name}_{pop}"][i],
+                    ht[f"exp_{annotation_name}_{pop}"][i],
+                )
+                # TODO: Should this have annotation_name in it?
+                full_pli_expr[f"pLI_{pop}"].append(pli_expr.pLI)
+                full_pli_expr[f"pRec_{pop}"].append(pli_expr.pRec)
+                full_pli_expr[f"pNull_{pop}"].append(pli_expr.pNull)
+
+    # Calculate pLI scores for all variants.
+    pli_expr = compute_pli(
+        ht, ht[f"obs_{annotation_name}"], ht[f"exp_{annotation_name}"]
+    )
+    full_pli_expr.update(
+        {f"{k}_{annotation_name}": v for k, v in list(pli_expr.items())}
+    )
+    return hl.struct(**full_pli_expr)
+
+
+def compute_pli(
+    ht: hl.Table, obs_expr: hl.expr.Int32Expression, exp_expr: hl.expr.Float32Expression
+) -> hl.expr.StructExpression:
+    """
+    Compute the pLI score using the observed and expected variant counts.
+
+    The output Table will include the following annotations:
+        - pLI - Probability of loss-of-function intolerance; probability that transcript
+          falls into distribution of haploinsufficient genes
+        - pNull - Probability that transcript falls into distribution of unconstrained
+          genes
+        - pRec - Probability that transcript falls into distribution of recessive genes
+
+    :param ht: Input Table.
+    :param obs_expr: Expression for the number of observed variants on each gene or
+        transcript in `ht`.
+    :param exp_expr: Expression for the number of expected variants on each gene or
+        transcript in `ht`.
+    :return: Table with pLI scores.
+    """
+    # Set up initial values.
+    last_pi = {"Null": 0, "Rec": 0, "LI": 0}
+    pi = {"Null": 1 / 3, "Rec": 1 / 3, "LI": 1 / 3}
+    # TODO: check these initial values and add a comment for where they come from.
+    expected_values = {"Null": 1, "Rec": 0.463, "LI": 0.089}
+
+    def get_pi(pi):
+        pi_expr = {
+            k: v * hl.dpois(obs_expr, exp_expr * expected_values[k])
+            for k, v in pi.items()
+        }
+        pi_expr["row_sum"] = hl.sum([pi_expr[k] for k in pi])
+        pi_expr = {
+            k: hl.or_missing(exp_expr > 0, pi_expr[k] / pi_expr["row_sum"])
+            for k, v in pi.items()
+        }
+        return pi_expr, ht.aggregate({k: hl.agg.mean(pi_expr[k]) for k in pi.keys()})
+
+    # Calculate pLI scores.
+    # TODO: should 0.001 be a parameter, and why is it 0.001?
```

Comment:
from page 34 of https://www.nature.com/articles/nature19057#Sec16 supplement:

"We use an expectation-maximization (EM) algorithm to find the MLE for π and Zi, treating π as the parameters and the Zi as the latent variables. We initialize the EM algorithm by setting 𝜋! = (1 3,1 3,1 3).......In the E-step, we evaluate the distribution of the latent variables (Zi) given the values of the parameters (π) from the previous iteration...... In the M-step, we update the parameters π with a new expectation taken under the distribution of the latent variables (Zi) computed in the M step.....We repeat these steps until the convergence criteria are met (𝜋!" changes by less than 0.001 from one iteration to the next)."